ubi_manifest/worker/tasks/content_audit.py

```diff
@@ -0,0 +1,146 @@
+    get_criteria_for_modules,
+    is_blacklisted,
+    keep_n_latest_modulemd_defaults,
+    keep_n_latest_modules,
+    keep_n_latest_rpms,
+    make_pulp_client,
+    parse_blacklist_config,
+)
+
+_LOG = logging.getLogger(__name__)
+MAX_WORKERS = int(os.getenv("UBI_MANIFEST_CONTENT_AUDIT_WORKERS", "8"))
+
+
+@app.task  # type: ignore [misc]  # ignore untyped decorator
+def content_audit_task() -> None:
+    """
+    This task checks that all available content is up-to-date, that whitelisted
+    content is present, and that blacklisted content is absent.
+    """
+
+    config_loaders = [UbiConfigLoader(url) for url in app.conf.content_config.values()]
+
+    with make_pulp_client(app.conf) as client:
+        for out_repo in client.search_repository(
+            Criteria.with_field("ubi_population", True)
+        ):
+            # get all relevant units currently on output repo
+            out_rpms = search_rpms([Criteria.true()], [out_repo]).result()
+            out_mds = search_modulemds([Criteria.true()], [out_repo]).result()
+            out_mdds = search_modulemd_defaults([Criteria.true()], [out_repo]).result()
+
+            seen_units = set()
+            output_whitelist = set()
+
+            for in_repo in client.search_repository(
+                Criteria.with_id(out_repo.population_sources)
+            ):
+                # get all corresponding units currently on input repo
+                in_rpms_ft = search_rpms(
+                    create_or_criteria(["name"], [(rpm.name,) for rpm in out_rpms]),
+                    [in_repo],
+                )
+                in_mds_ft = search_modulemds(
+                    get_criteria_for_modules(list(out_mds)), [in_repo]
+                )
+                in_mdds_ft = search_modulemd_defaults(
+                    get_criteria_for_modules(list(out_mdds)), [in_repo]
+                )
+
+                # check that all content is up-to-date
+                in_rpms = list(in_rpms_ft.result())
+                keep_n_latest_rpms(in_rpms)
+                for in_rpm in in_rpms:
+                    for out_rpm in out_rpms:
+                        if in_rpm.name == out_rpm.name:
+                            _compare_versions(out_repo.id, in_rpm, out_rpm)
+                            seen_units.add(in_rpm)
+                in_mds = list(in_mds_ft.result())
+                keep_n_latest_modules(in_mds)
+                for in_md in in_mds:
+                    for out_md in out_mds:
+                        if in_md.name + in_md.stream == out_md.name + out_md.stream:
+                            _compare_versions(out_repo.id, in_md, out_md)
+                            seen_units.add(in_md)
+                in_mdds = list(in_mdds_ft.result())
+                keep_n_latest_modulemd_defaults(in_mdds)
+                for in_mdd in in_mdds:
+                    for out_mdd in out_mdds:
+                        if in_mdd.name + in_mdd.stream == out_mdd.name + out_mdd.stream:
```

Comment:
This may 'soon' be problematic as well because in rhsm-pulp there are already  `modulemd_defaults` with undefined stream, so I guess if e.g. `out_mdd.stream` is `None` it's going to crash. Do you think you can re-write this so it's a bit future-proof?